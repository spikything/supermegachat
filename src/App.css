body {
  background-color: rgb(0, 0, 0);
}

.App {
  text-align: center;
  max-width: 720px;
  margin: 0;
}

.App header {
  background-color: rgb(58, 58, 58);
  border-radius: 0px 20px 20px 0px;
  height: 50px;
  min-height: 50px;
  color: white;
  position: fixed;
  width: 100%;
  max-width: 720px;
  top: 0;
  display: flex;
  align-items: center;
  justify-content: space-between;
  z-index: 99;
  padding: 0px 0px;
  padding-left: 0px;
  box-sizing: border-box;
}

h2 {
  color: #acacac;
  margin-left: -25px;
}

.App section {
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-height: 100vh;
  background-color: rgb(0, 0, 0);
}

main {
  padding: 10px;
  height: 100vh;
  margin: 10vh 0 10vh;
  overflow-y: scroll;
  display: flex;
  flex-direction: column;
}

main::-webkit-scrollbar {
  width: 0.25rem;
}

main::-webkit-scrollbar-track {
  background: #1e1e24;
}

main::-webkit-scrollbar-thumb {
  background: #3f3dad;
}

form {
  height: 50px;
  min-height: 50px;
  border-radius: 0px 20px 20px 0px;
  position: fixed;
  bottom: 0;
  background-color: rgb(58, 58, 58);
  width: 100%;
  max-width: 720px;
  display: flex;
  font-size: 1.5rem;
  z-index: 999;
}

button {
  background-color: rgb(206, 9, 58);
  width: 30%;
  border: none;
  color: white;
  border-radius: 0px 20px 20px 0px;
  padding: 10px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  cursor: pointer;
  font-size: 1.25rem;
  height: 100%;
}

form button {
  width: 30%;
  padding-left: -20px;
  padding-right: 15px;
}

input {
  background: rgb(58, 58, 58);
  color: white;
  line-height: 1.5;
  width: 100%;
  font-size: 1.5rem;
  outline: none;
  border: none;
  padding: 0 10px;
  margin-left: 5px;
}

.sign-in {
  width: 55%;
  max-width: 55%;
  align-self: center;
  border-radius: 20px 20px 20px 0px;
}

button:focus {
  outline: none;
}

button:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

ul, li {
  text-align: left;
  list-style: none;
}

p {
  max-width: 70vw;
  margin-bottom: 5px;
  margin-top: 5px;
  line-height: 20px;
  padding: 10px 20px;
  padding-left: 16px;
  padding-right: 16px;
  border-radius: 20px 20px 20px 0px;
  position: relative;
  color: white;
  text-align: left;
}

.messageWindow {
  background-color: #000000;
  height: calc(100vh - 100px);
  overflow-x: hidden;
  overflow-y: scroll;
}

.message {
  display: flex;
  align-items: center;
}

.sent,
.read {
  flex-direction: row-reverse;
}

.sent p,
.read p {
  color: #b1c9db;
  background: #135c91;
  border-radius: 20px 20px 0px 20px;
  align-self: flex-end;
}
.received p {
  background: #cedce6;
  color: black;
}
.read p {
  color: white;
  background: #0d81d4;
}

img {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  margin: 2px 5px;
  z-index: 9;
}
