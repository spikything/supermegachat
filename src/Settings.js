const Settings = {
  LOGGING_ENABLED: false,
  MAX_MESSAGES: 100,
  SOFT_KEYBOARD_OPEN_DELAY: 50,
  SYSTEM_MESSAGES_ENABLED: false,
  FIREBASE_INIT_OBJECT: {
    apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
    authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_FIREBASE_DATABASE_URL,
    projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
    storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_FIREBASE_APP_ID,
    measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID,
  }
};

export default Settings;
