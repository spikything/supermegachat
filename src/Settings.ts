const Settings = {

  LOGGING_ENABLED: false,
  MAX_MESSAGES: 100,
  SOFT_KEYBOARD_OPEN_DELAY: 50,
  SYSTEM_MESSAGES_ENABLED: false,
  USER_PASS_AUTH_ENABLED: false,
  AUTH_EMULATOR: "http://localhost:9099",
  FIREBASE_EMULATOR: "localhost:8080",
  USE_EMULATOR_WHEN_LOCAL: false, // if you set this to true, ensure the firebase emulators are running for testing (firebase emulators:start)
  TEST_USER: process.env.TEST_USER,
  TEST_PASS: process.env.TEST_PASS,

  FIREBASE_INIT_OBJECT: {
    apiKey: process.env.REACT_APP_FIREBASE_API_KEY,
    authDomain: process.env.REACT_APP_FIREBASE_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_FIREBASE_DATABASE_URL,
    projectId: process.env.REACT_APP_FIREBASE_PROJECT_ID,
    storageBucket: process.env.REACT_APP_FIREBASE_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_FIREBASE_APP_ID,
    measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENT_ID,
  },

  get isLocal() {
    return window.location.hostname === "localhost"
  }

};

export default Settings;
