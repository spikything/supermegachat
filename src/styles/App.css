@import './themes/Light.css';
@import './themes/Dark.css';

body {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: var(--bg, #111);
  transition: background-color 0.4s ease-in-out;
}

.App {
  text-align: center;
  max-width: 720px;
  margin: 0;
}

.App header {
  background-color: var(--header);
  border-radius: 0px 20px 20px 0px;
  height: 50px;
  min-height: 50px;
  color: white;
  position: fixed;
  width: 100%;
  max-width: 720px;
  top: 0;
  display: flex;
  align-items: center;
  /* justify-content:space-between; */
  z-index: 99;
  padding: 0px 0px;
  padding-left: 0px;
  box-sizing: border-box;
  -webkit-box-shadow: 0px 0px 8px 0px rgba(0,0,0,0.95); 
  box-shadow: 0px 0px 8px 0px rgba(0,0,0,0.95);
}

h1 {
  color: var(--text);
  margin: 0 5%;
}

header h2 {
  color: var(--text);
  margin-left: 5px;
}

header button {
  right: 0px;
  position: absolute;
}

.App section {
  display: flex;
  flex-direction: column;
  justify-content: center;
  min-height: 100vh;
  /* background-color: rgb(0, 0, 0); */
}

h1.loading {
  filter: grayscale(100);
}

main {
  padding: 10px;
  height: 100vh;
  margin: 10vh 0 10vh;
  overflow-y: scroll;
  display: flex;
  flex-direction: column;
}

main::-webkit-scrollbar {
  width: 0.25rem;
}

main::-webkit-scrollbar-track {
  background: #1e1e24;
}

main::-webkit-scrollbar-thumb {
  background: #3f3dad;
}

form {
  height: 50px;
  min-height: 50px;
  border-radius: 0px 20px 20px 0px;
  position: fixed;
  bottom: 0;
  background-color: var(--header);
  width: 100%;
  max-width: 720px;
  display: flex;
  font-size: 1.5rem;
  z-index: 999;
  -webkit-box-shadow: 0px 0px 8px 0px rgba(0,0,0,0.95); 
  box-shadow: 0px 0px 8px 0px rgba(0,0,0,0.95);
}

button {
  background-color: var(--button);
  width: 30%;
  border: none;
  color: var(--button-text);
  border-radius: 0px 20px 20px 0px;
  padding: 10px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  cursor: pointer;
  font-size: 1.25rem;
  height: 100%;
}

form button {
  width: 30%;
  padding-left: -20px;
  padding-right: 15px;
}

input {
  background: var(--header);
  color: var(--text);
  line-height: 1.5;
  width: 100%;
  font-size: 1.5rem;
  outline: none;
  border: none;
  padding: 0 10px;
  margin-left: 5px;
}

.sign-in {
  width: 55%;
  max-width: 55%;
  align-self: center;
  border-radius: 20px 20px 20px 0px;
}

button:focus {
  outline: none;
}

button:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

ul, li {
  text-align: left;
  list-style: none;
}

p {
  -webkit-user-select: text;
  -moz-user-select: text;
  -ms-user-select: text;
  user-select: text;
  min-width: 1vw;
  max-width: 70vw;
  margin-bottom: 5px;
  margin-top: 5px;
  margin-right: 0px;
  line-height: 20px;
  padding: 10px;
  border-radius: 20px 20px 20px 0px;
  position: relative;
  color: white;
  text-align: left;
}

.messageWindow {
  /* background-color: #000000; */
  height: calc(100vh - 100px);
  overflow-x: hidden;
  overflow-y: scroll;
}

.message {
  display: flex;
  align-items: center;
}

.sent,
.read {
  flex-direction: row-reverse;
  padding: 0px;
}

.sent p,
.read p {
  color: #b1c9db;
  background: #135c91;
  border-radius: 20px 20px 0px 20px;
  align-self: flex-end;
}
.received p {
  background: #cedce6;
  color: black;
}
.read p {
  color: white;
  background: #0d81d4;
}

.fromsystem p {
  color: #999cb4;
  width: 100vw;
  max-width: 100vw;
  text-align: center;
  background: none;
}

.avatar {
  width: 40px;
  height: 40px;
  border-radius: 50%;
  margin: 2px 5px;
  z-index: 9;
}

.messageimage {
  cursor: pointer;
}

img {
  filter: grayscale(var(--image-filter)) brightness(calc(100% - (var(--image-filter) / 2)));
  -webkit-filter: grayscale(var(--image-filter)) brightness(calc(100% - (var(--image-filter) / 2)));
  transition: filter 1s;
}

.logo {
  width: 50%;
  height: auto;
  border-radius: 0;
  margin: 10%;
  align-self: center;
}

img.seen {
  width: 14px;
  height: 14px;
  align-self: flex-end;
  margin: 0px;
  margin-bottom: 3px;
  padding: 0;
}

.sent img.seen,
.received img.seen,
.fromsystem img.seen {
  display: none;
}

.fromsystem img {
  display: none;
}
