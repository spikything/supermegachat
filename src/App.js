import React, { useRef, useEffect } from 'react';
import useSound from 'use-sound';
import boopSound from './menu-open.mp3';
import './App.css';

import firebase from 'firebase/app';
import 'firebase/firestore';
import 'firebase/auth';

import {useAuthState} from 'react-firebase-hooks/auth';
import {useCollectionData} from 'react-firebase-hooks/firestore';

firebase.initializeApp({
  apiKey: "AIzaSyACHocPkDUzh59OfnqXtfx0SkvaKHEfrmA",
  authDomain: "superchat-11bdc.firebaseapp.com",
  databaseURL: "https://superchat-11bdc.firebaseio.com",
  projectId: "superchat-11bdc",
  storageBucket: "superchat-11bdc.appspot.com",
  messagingSenderId: "206674897831",
  appId: "1:206674897831:web:a33e919945e6ed554b6482",
  measurementId: "G-201RN4JJRW"
});

const auth = firebase.auth();
const firestore = firebase.firestore();

function App() {

  // Google Authentication
  const [user] = useAuthState(auth);

  // Disable pinch zooming
  const useDisablePinchZoomEffect = () => {
    useEffect(() => {
      const disablePinchZoom = (e) => {
        if (e.touches.length > 1) {
          e.preventDefault()
        }
      }
      document.addEventListener("touchmove", disablePinchZoom, { passive: false })
      return () => {
        document.removeEventListener("touchmove", disablePinchZoom)
      }
    }, [])
  }
  useDisablePinchZoomEffect();

  // CONSTRUCT APP
  return (
    <div className="App">

      { user ? <Header /> : null }

      <section>
        { user ? <ChatRoom /> : <SignIn /> }
      </section>

    </div>
  );
}

function SignIn () {

  const signInWithGoogle = () => {
    const provider = new firebase.auth.GoogleAuthProvider();
    auth.signInWithPopup(provider);
  }

  return (
    <button className="sign-in" onClick={signInWithGoogle}>Sign into MegaChat with Google</button>
  )
}

function Header () {
  return auth.currentUser && (
    <header>
      <img src={ auth.currentUser.photoURL || 'https://lh3.googleusercontent.com/fife/ABSRlIphM2hawV2AzwI5CEphA9pmHalAjD5pYdiLKYT2PMULhdUKQ5d6vxakdSBF1CSq8XoG0d0a9RGJp3iwAVYizt7sleGdBp5ltpG03Q7SW8pi3OEmaIqsar75CU9rIj5g8S9hPR6BjCNWaIBuptJDmzVHR4yE43fuWx1VRg8ITMXg7__PuodYGemf9_J0UYTBuRRr8hxJc4kT6nleESBisdIjmupgaj7eXU0Cq6DenzcwS9TDOmQM5dzDIlR4goowjejdREdtUJff3neOf8wt5asWiIUoTsYMMJRJAncPdv7CFLZ79AEtxJa2ZBNv9DEXY-E50Cdg8IUS4-cVrIXO2uA52AzKhylu0c9iK9gg1cr3e0JLRU91-jq3oKk8I-6rD5wVK6bm4OK7jQ7Qg-Zzee1_CjFBFImXUddZhetObZniSB1RtE1nDT_RUoqsKh-LSV5cYu0jUHH62f5lATcM4d8sJFqkR_qWXxTnLcv0xoVaZSHCqP9L7BxVm_XVA1J9BGSfy21cnEr7k-Zmhcu_HVazv1l74J7cJWRJiUIn9goxy2bWWi7YZWVyFllMoLnlhSLjLrCcnp3bx901CnXE38ZTbvbtnaZhJDthk8tQ44RhjB-Fnbpjm7IkNuKd0iBDlbSHXD0RWvphCfSbZKVcumYLxhLwoNfkhZjpEdnLceSD8aUNrL0JA_H2_dNgVA5pvdh9d1Vw0aNBxc8LPFDKSXiTE5HQv4LT=s5184-w5184-h2916-no' } alt='' />
      <h2>{ auth.currentUser.displayName }</h2>
    <SignOut />
  </header>
  )
}

function SignOut () {
  return auth.currentUser && (
    <button onClick={() => auth.signOut()}>Sign out</button>
  )
}

function ChatRoom () {

  const dummy = useRef(undefined);
  const formRef = useRef(undefined);

  const messagesRef = firestore.collection('messages');
  const query = messagesRef.orderBy('createdAt', 'desc').limit(99);
  const [messages] = useCollectionData(query, {idField: 'id'});
  // const [formValue, setFormValue] = useState('');

  /*
  const onFormChange = (e) => {
    setFormValue(e.target.value);
    // clearTimeout(window.$scrollUpdateTime);
    // document.getElementById('messageBottom').scrollIntoView({ behavior: 'auto' });
  }
  */

  const sendMessage = async(e) => {
    e.preventDefault();

    const textinput = document.getElementById('textinput');
    let formValue = textinput.value;

    if (formValue === '') return;

    const { uid, photoURL, displayName } = auth.currentUser;

    await messagesRef.add({
	  text: formValue,
      createdAt: firebase.firestore.FieldValue.serverTimestamp(),
      uid,
      displayName,
      photoURL
    });

    // setFormValue('');
    textinput.value = '';
    //dummy.current.scrollIntoView({ behavior: 'smooth' });
  }

  // useEffect(() => {
  //   dummy.current.scrollIntoView({ behavior: "smooth" })
  // }, [])

  // TRYING USEEFFECT WITH TIMEOUT
  // useEffect(() => {
  //   clearTimeout(window.$scrollUpdateTime);
  //   window.$scrollUpdateTime = setTimeout(() => { document.getElementById('messageBottom').scrollIntoView({ behavior: 'smooth' }); }, 100);
  // }, []);

  return (
    <>
      <div className={`messageWindow`}>
        {messages && messages.reverse().map(msg => <ChatMessage key={msg.id} message={msg} /> )}

        <div id="messageBottom" ref={dummy}></div>
      </div>

      <form ref={formRef} onSubmit={sendMessage}>
        <input id="textinput" placeholder="Type here..." maxLength="500" autoComplete="off" />
        <button type="submit">SEND</button>
      </form>

    </>
  )
}

function ChatMessage (props) {
  // The dummy ref trick doesn't seem to respond to all incoming data. So use an environment variable to update scrolling once after a timeout
  // const dummy = useRef();

  const [playBoop] = useSound(boopSound);

  useEffect(() => {
    // clearTimeout(window.$scrollUpdateTime);
    // window.$scrollUpdateTime = setTimeout(() => { document.getElementById('messageBottom').scrollIntoView({ behavior: 'smooth' }); }, 100);
    playBoop();
  }, [playBoop]);

  const scrollToBottom = () => {
    clearTimeout(window.$scrollUpdateTime);
    window.$scrollUpdateTime = setTimeout(() => { document.getElementById('messageBottom').scrollIntoView({ behavior: 'smooth' }); }, 100);
  }

  const { text, uid, photoURL, createdAt, displayName } = props.message;
  const messageClass = uid === auth.currentUser.uid ? 'sent' : 'received';

  return (
    <div className={`message ${messageClass}`}>
      <img src={ photoURL } alt='' onLoad={scrollToBottom} onClick={ () => alert((displayName || uid) + '\n\nSent: ' + createdAt.toDate()) } />
      <p>{text}</p>
    </div>
  );
}

export default App;
