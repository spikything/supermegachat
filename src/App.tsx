import React, { useEffect } from "react";
import { useAuthState } from "react-firebase-hooks/auth";
import firebase from "firebase/app";
import "firebase/firestore";
import "firebase/auth";
import Settings from "./Settings";
import ChatRoom from "./components/ChatRoom";
import Header from "./components/Header";
import SignIn from "./components/SignIn";
import "./App.css";

// Firebase initilisation
const fsapp = firebase.initializeApp(Settings.FIREBASE_INIT_OBJECT);
if (Settings.isLocal && Settings.USE_EMULATOR_WHEN_LOCAL) {
  fsapp.auth().useEmulator(Settings.AUTH_EMULATOR);
  fsapp.firestore().settings({
    host: Settings.FIREBASE_EMULATOR,
    ssl: false,
  });
}

const auth = firebase.auth();
const firestore = firebase.firestore();

function App() {

  // Google Firebase Authentication
  const [user] = useAuthState(auth);

  // Setup
  useEffect(() => {

    // Disable pinch zooming
    const disablePinchZoom = (evt:TouchEvent) => {
      if (evt.touches.length > 1) {
        evt.preventDefault();
      }
    };
    document.addEventListener("touchmove", disablePinchZoom, { passive: false });

    // Component teardown callback
    return () => {
      document.removeEventListener("touchmove", disablePinchZoom);
    };
  }, []);

  // The component JSX to render
  return (
    <div className="App">
      {user ? <Header auth={auth} /> : null}

      <section>
        {user ? (
          <ChatRoom
            Settings={Settings}
            auth={auth}
            firestore={firestore}
          />
        ) : (
          <SignIn 
            auth={auth} 
          />
        )}
      </section>
    </div>
  );
}

export default App;
