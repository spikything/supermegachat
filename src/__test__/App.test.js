import { act, render, screen } from '@testing-library/react';
import App from '../App';
import Settings from '../Settings';

// This script provides basic smoke testing of the app. 
// More extensive E2E testing is done with the WebDriver (see webdriver folder)

describe('App smoke test', () => {
  test('Initial render without error', async () => {
    await act(async () => {

      render(<App />);

      const linkElement = screen.getByText(Settings.SIGN_IN_LABEL);
      expect(linkElement).toBeInTheDocument();

    });
  });
});
