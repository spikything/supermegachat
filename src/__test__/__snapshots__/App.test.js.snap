// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <section>
          <h1
            style="opacity: 0; transform: translateX(100px) translateZ(0);"
          >
            Welcome to MegaChat
          </h1>
          <img
            alt="logo"
            class="logo"
            src="logo512.png"
            style="opacity: 0; transform: translateX(-100px) translateZ(0);"
          />
          <button
            class="sign-in"
            style="opacity: 0; transform: translateY(100px) translateZ(0);"
          >
            Sign in with Google
          </button>
          <input
            class="checkbox"
            id="checkbox"
            type="checkbox"
          />
          <label
            class="label signin"
            for="checkbox"
            style="opacity: 0; transform: translateX(150px) scale(1.5) translateZ(0);"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-moon "
              data-icon="moon"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 256c0-123.8 100.3-224 223.8-224c11.36 0 29.7 1.668 40.9 3.746c9.616 1.777 11.75 14.63 3.279 19.44C245 86.5 211.2 144.6 211.2 207.8c0 109.7 99.71 193 208.3 172.3c9.561-1.805 16.28 9.324 10.11 16.95C387.9 448.6 324.8 480 255.8 480C132.1 480 32 379.6 32 256z"
                fill="currentColor"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-sun "
              data-icon="sun"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 159.1c-53.02 0-95.1 42.98-95.1 95.1S202.1 351.1 256 351.1s95.1-42.98 95.1-95.1S309 159.1 256 159.1zM509.3 347L446.1 255.1l63.15-91.01c6.332-9.125 1.104-21.74-9.826-23.72l-109-19.7l-19.7-109c-1.975-10.93-14.59-16.16-23.72-9.824L256 65.89L164.1 2.736c-9.125-6.332-21.74-1.107-23.72 9.824L121.6 121.6L12.56 141.3C1.633 143.2-3.596 155.9 2.736 164.1L65.89 256l-63.15 91.01c-6.332 9.125-1.105 21.74 9.824 23.72l109 19.7l19.7 109c1.975 10.93 14.59 16.16 23.72 9.824L256 446.1l91.01 63.15c9.127 6.334 21.75 1.107 23.72-9.822l19.7-109l109-19.7C510.4 368.8 515.6 356.1 509.3 347zM256 383.1c-70.69 0-127.1-57.31-127.1-127.1c0-70.69 57.31-127.1 127.1-127.1s127.1 57.3 127.1 127.1C383.1 326.7 326.7 383.1 256 383.1z"
                fill="currentColor"
              />
            </svg>
            <div
              class="ball"
            />
          </label>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <section>
        <h1
          style="opacity: 0; transform: translateX(100px) translateZ(0);"
        >
          Welcome to MegaChat
        </h1>
        <img
          alt="logo"
          class="logo"
          src="logo512.png"
          style="opacity: 0; transform: translateX(-100px) translateZ(0);"
        />
        <button
          class="sign-in"
          style="opacity: 0; transform: translateY(100px) translateZ(0);"
        >
          Sign in with Google
        </button>
        <input
          class="checkbox"
          id="checkbox"
          type="checkbox"
        />
        <label
          class="label signin"
          for="checkbox"
          style="opacity: 0; transform: translateX(150px) scale(1.5) translateZ(0);"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-moon "
            data-icon="moon"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 256c0-123.8 100.3-224 223.8-224c11.36 0 29.7 1.668 40.9 3.746c9.616 1.777 11.75 14.63 3.279 19.44C245 86.5 211.2 144.6 211.2 207.8c0 109.7 99.71 193 208.3 172.3c9.561-1.805 16.28 9.324 10.11 16.95C387.9 448.6 324.8 480 255.8 480C132.1 480 32 379.6 32 256z"
              fill="currentColor"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-sun "
            data-icon="sun"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 159.1c-53.02 0-95.1 42.98-95.1 95.1S202.1 351.1 256 351.1s95.1-42.98 95.1-95.1S309 159.1 256 159.1zM509.3 347L446.1 255.1l63.15-91.01c6.332-9.125 1.104-21.74-9.826-23.72l-109-19.7l-19.7-109c-1.975-10.93-14.59-16.16-23.72-9.824L256 65.89L164.1 2.736c-9.125-6.332-21.74-1.107-23.72 9.824L121.6 121.6L12.56 141.3C1.633 143.2-3.596 155.9 2.736 164.1L65.89 256l-63.15 91.01c-6.332 9.125-1.105 21.74 9.824 23.72l109 19.7l19.7 109c1.975 10.93 14.59 16.16 23.72 9.824L256 446.1l91.01 63.15c9.127 6.334 21.75 1.107 23.72-9.822l19.7-109l109-19.7C510.4 368.8 515.6 356.1 509.3 347zM256 383.1c-70.69 0-127.1-57.31-127.1-127.1c0-70.69 57.31-127.1 127.1-127.1s127.1 57.3 127.1 127.1C383.1 326.7 326.7 383.1 256 383.1z"
              fill="currentColor"
            />
          </svg>
          <div
            class="ball"
          />
        </label>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
